core/corehttp/gateway_handler.go

```diff
@@ -342,6 +342,15 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 		BackLink: backLink,
 		Hash:     hash,
 	}
+
+	// See statusResponseWriter.WriteHeader
+	// and https://github.com/ipfs/go-ipfs/issues/7164
+	// Note: this needs to occur before listingTemplate.Execute otherwise we get
+	// superfluous response.WriteHeader call from prometheus/client_golang
+	if w.Header().Get("Location") != "" {
+		w.WriteHeader(http.StatusMovedPermanently)
```

Comment:
@Stebalien Do we need to return directory listing payload with the redirect?
I think it would be safer to `return` here, right after we write the status code header.

Reply:
I agree. Directory listings are for browsers.

Reply:
@Stebalien updated this PR to include mentioned `return`